### PR TITLE
mlxrunner: unify and tune speculative decoding

### PR DESCRIPTION
**Pull request by jessegross**
MTP previously drafted a fixed number of tokens per round — a constant by default, or an opt-in heuristic that grew toward a cap on acceptance. Neither targeted throughput: a constant depth couldn't adapt to the target, and the heuristic maximized accepted tokens per step, so on a steep forward curve it kept drafting deeper while tokens per second fell, ending up slower than plain decode.

This replaces both with a controller that picks the draft length to maximize committed tokens per second, from measured per-position acceptance and forward cost. It backs off to plain decode when no draft pays and persists across requests, so each one starts at the proven-out depth instead of re-ramping from shallow.

This change also:

- **Unifies the decode paths.** Greedy, sampled, and serial paths and per-model draft branches collapse into one engine and a single drafter interface that every draft head implements.
- **Cuts per-step overhead.** Two costs that capped speculation's benefit go away: the current token and its draft validation share one target forward, and each round resolves in one host sync instead of two.
- **Adds draft-cache infrastructure for Qwen3.6 MTP.** A draft head can maintain its own KV caches, prefix-cached across requests, instead of reusing the target's.

The `OLLAMA_MLX_MTP_*` environment variables are removed along with the old schedule.

### CODE REVIEW

File: x/mlxrunner/runner.go [hunk 1]
```diff
@@ -78,7 +80,10 @@ func (r *Runner) Load(modelName string) error {
 		if err := draft.LoadWeights(tensors); err != nil {
 			return err
 		}
-		r.Draft = draft
+		draftModel = draft
+	} else if sd, ok := m.(base.SelfDraft); ok {
+		// Inline draft head: already loaded with the target; nil if none shipped.
+		draftModel = sd.SelfDraft()
```
Comment (pdevine) [inline x/mlxrunner/runner.go:86]: I'm assuming this is for Qwen3Next architectures.

File: x/mlxrunner/pipeline.go [hunk 2]
```diff
@@ -132,86 +161,97 @@ func (r *Runner) TextGenerationPipeline(ctx context.Context, request Request) er
...
 		}
 
-		mlx.Unpin(sample.Arrays()...)
-		sample, nextSample = nextSample, sampler.Result{}
+		if done {
+			break
+		}
 
-		if i%256 == 0 {
+		if generated%256 == 0 {
```
Comment (pdevine) [inline x/mlxrunner/pipeline.go]: this should probably be a constant

File: x/mlxrunner/speculate.go [hunk 3]
```diff
@@ -0,0 +1,531 @@
...
+// bind computes the draft/target cache partition the first time the persistent
+// caches exist; later requests reuse the same slice, so it runs once.
+func (s *speculation) bind(caches []cache.Cache) {
+	if s.caches != nil {
+		if !slices.Equal(s.caches, caches) {
+			panic("speculation: cache slice changed between requests")
+		}
+		return
+	}
+	draftKV := s.draft.DraftCaches(caches)
```
Comment (pdevine) [inline x/mlxrunner/speculate.go:82]: I'm assuming this gets used in the future w/ qwen35? Gemma4 looks like it doesn't use the draft cache.

File: x/mlxrunner/mtp.go [hunk 4]
```diff
@@ -1,876 +1,256 @@
...
+			// head stays at the last committed slot every step, re-attending
+			// the committed prefix read-only ("single-position").
+			pos := d.frontier - 1
+			if len(d.spec.draftKV) > 0 {
+				pos = d.frontier - 1 + i
+			}
+			hidden, projected = d.spec.draft.Draft(&batch.Batch{
+				InputIDs:     lastToken,
+				SeqOffsets:   []int32{int32(pos)},
+				SeqQueryLens: []int32{1},
```
Comment (pdevine) [inline x/mlxrunner/mtp.go:228]: I'm assuming this is for auto-regressive models and would change if this was diffusion based? Or would that map on a different way.

File: x/models/gemma4/assistant.go [hunk 5]
```diff
@@ -267,26 +269,42 @@ func (m *AssistantModel) precomputeScaledWeights() {
...
-func (m *AssistantModel) Draft(inputsEmbeds *mlx.Array, position int32, caches []cache.Cache) (logits, hidden *mlx.Array) {
+// DraftCaches returns nil: the assistant keeps no KV and re-attends the
+// target's caches read-only each step.
+func (m *AssistantModel) DraftCaches([]cache.Cache) []cache.Cache { return nil }
+
+// Draft is single-position: the head drafts every token as if it sat at the
+// last target-seen position, so it anchors RoPE and the mask there — from the
+// non-moving target full-attention cache — regardless of the advancing offset
+// in b. The input fuses the target's scaled token embedding with b.Hidden.
+func (m *AssistantModel) Draft(b *batch.Batch, caches []cache.Cache) (logits, projected *mlx.Array) {
```
Comment (pdevine) [inline x/models/gemma4/assistant.go]: Did you want to call `logits` as `hidden` here similar to how you're naming `projected`?

File: x/mlxrunner/mtp.go [hunk 6]
```diff
@@ -1,876 +1,256 @@
...
-	}
-	return mlx.Concatenate(draftTokens, 1)
-}
-
-func (r *Runner) generateMTPDraftCandidates(draft base.MTPDraftModel, target base.MTPEmbeddingModel, token *mlx.Array, hidden *mlx.Array, caches []cache.Cache, position int32, maxDraft int) *mtpDraftCandidates {
-	if maxDraft <= 0 {
-		return nil
+	if len(d.spec.draftKV) > 0 {
+		d.settle(current)
+		if d.heldHidden == nil {
```
Comment (pdevine) [inline x/mlxrunner/mtp.go:199]: I don't think this will execute right now (since there's no draftKV), but what is d.heldHidden being used for here?

File: x/mlxrunner/mtp.go [hunk 7]
```diff
@@ -1,876 +1,256 @@
...
+		// pairing resumes from their absolute offset.
+		d.committedDraftOffset = s.draftKV[0].Offset()
+		d.frontier = d.committedDraftOffset
+	}
+	return d
+}
+
+func (d *mtpDrafter) committed(tokens, hiddens *mlx.Array, position int) {
+	n := tokens.Dim(1)
+	if len(d.spec.draftKV) > 0 {
```
Comment (pdevine) [inline x/mlxrunner/mtp.go:59]: I'm assuming all of this gets skipped until we do the qwen3.5 mtp.

File: x/mlxrunner/mtp.go [hunk 8]
```diff
@@ -1,876 +1,256 @@
...
+	if d.pendingCount >= mtpPendingFlushTokens {
+		d.flush()
+	}
+}
+
+// flush writes the pending pairs to the draft caches in one head forward,
+// dropping speculative entries past the committed range first and holding
+// the last row's logits and projected hidden for the next proposal chain.
+func (d *mtpDrafter) flush() {
+	if len(d.pendingTokens) == 0 {
```
Comment (pdevine) [inline x/mlxrunner/mtp.go:138]: This was a little bit hard to trace through, but for now it looks like the function will return immediately since d.pendingTokens won't get set unless d.spec.draftKV is set.

File: x/mlxrunner/speculate.go [hunk 9]
```diff
@@ -0,0 +1,531 @@
...
+	if s == nil {
+		return nil
+	}
+	s.bind(caches)
+	d := newMTPDrafter(s)
+
+	// Logprobs are not yet supported, so a logprobs request keeps a speculationSession
+	// only to maintain a draft cache in lockstep (permanently parked); a
+	// draftless head has nothing to maintain and decodes plainly.
+	opts := request.SamplerOpts
```
Comment (pdevine) [inline x/mlxrunner/speculate.go:130]: ok, this tripped me up a little bit:
For gemma4 MTP -> return nil if either logprobs or toplogprobs are specified
For qwen35 MTP (not supported yet) -> return the speculationSession

Would it make sense to just park gemma4 MTP w/ a speculationSession if logprobs/toplogprobs was specified? (is that possible?)

File: x/mlxrunner/speculate_depth.go [hunk 10]
```diff
@@ -0,0 +1,304 @@
...
+		if ev := c.committed(n) / c.cost.cost(n); ev > bestEV {
+			best, bestEV = n, ev
+		}
+	}
+	return best
+}
+
+// costEWMAAlpha weights the newest cost sample. Fixed-depth cost is low-variance,
+// so a responsive alpha converges in a few visits while smoothing scheduler jitter.
+const costEWMAAlpha = 0.3
```
Comment (pdevine) [inline x/mlxrunner/speculate_depth.go:123]: How do you come up with the moving average alpha and the clamp fraction? Just best guesses?

File: x/mlxrunner/speculate.go [hunk 11]
```diff
@@ -0,0 +1,531 @@
...
+func (s *speculationSession) endRound(drafted, accepted, observed int) {
+	s.roundDrafts = drafted
+	s.stats.iterations++
+	s.stats.drafted += drafted
+	s.stats.accepted += accepted
+	if s.enabled {
+		if observed > 0 {
+			s.spec.depth.acc.observe(observed, accepted)
+		}
+		s.limit = s.spec.depth.next()
```
Comment (pdevine) [inline x/mlxrunner/speculate.go:168]: I think this might have a subtle small bug in it. Because this is the final round and it calls next() the probe depth never gets used (but it sets pending = true). Then open() on the next round calls depth.next() and the probe ends up backing off.

**Review by pdevine — APPROVED** [review]
OK, overall looks really good. Most of my comments are just about only having Gemma4 right now; I'll dive deeper when we add the Qwen3Next MTP work.

File: x/mlxrunner/runner.go (hunk 1 above)
Comment (jessegross) [reply x/mlxrunner/runner.go:86]: Yes, that's right. It's also used by most other non-Gemma4 MTP architectures (Qwen, DeepSeek, GLM).

File: x/mlxrunner/pipeline.go (hunk 2 above)
Comment (jessegross) [reply x/mlxrunner/pipeline.go]: This is pre-existing code moving around but agreed and folded in.

File: x/mlxrunner/speculate.go (hunk 3 above)
Comment (jessegross) [reply x/mlxrunner/speculate.go:82]: Yes and same comment about this also being common to most non-Gemma MTP architectures.

File: x/mlxrunner/mtp.go (hunk 4 above)
Comment (jessegross) [reply x/mlxrunner/mtp.go:228]: The mtp.go file implements the `drafter` interface. A diffusion based drafter could have a separate implementation of that and still plug into all of the runner, verification, depth estimation, etc. in the same way.

File: x/models/gemma4/assistant.go (hunk 5 above)
Comment (jessegross) [reply x/models/gemma4/assistant.go]: Yes, I should have updated the implementation when changing the interface - fixed.

File: x/mlxrunner/mtp.go (hunk 6 above)
Comment (jessegross) [reply x/mlxrunner/mtp.go:199]: Correct, it doesn't execute with gemma4.

For models with draft caches, we need to keep them up to date with previous committed tokens, prefill or even when we have disabled MTP (in case we turn it on again). This is done with a batched pass over the draft model. That batch pass will also produce a hidden value as an output, which we can use later when we predict the first draft rather than doing a separate decode pass. `heldHidden` holds that.

File: x/mlxrunner/mtp.go (hunk 7 above)
Comment (jessegross) [reply x/mlxrunner/mtp.go:59]: Yes

File: x/mlxrunner/mtp.go (hunk 8 above)
Comment (jessegross) [reply x/mlxrunner/mtp.go:138]: Yes, that's correct. Related to the `heldHidden` question above, `pendingTokens` are tokens that are waiting to be stored in the draft caches - we batch them together until there is a good sized group or we need to flush.

File: x/mlxrunner/speculate.go (hunk 9 above)
Comment (jessegross) [reply x/mlxrunner/speculate.go:130]: Yes, it is a little conceptually simpler for all MTP models to have a `speculationSession` and just permanently park if there is something that prevents it from doing MTP.

For qwen this is necessary because logprobs can be enabled on a per-request basis so we need to keep the draft caches up to date - otherwise it is impossible to speculate with a potential future non-logprobs request. Gemma4 doesn't have draft caches so it doesn't have this issue but it's fine to follow the same pattern. Changed now.

We should support logprobs with MTP in the future, so this case should go away, but for now it will behave as expected, just without the speed benefit of MTP.

File: x/mlxrunner/speculate_depth.go (hunk 10 above)
Comment (jessegross) [reply x/mlxrunner/speculate_depth.go:123]: They were initially best guesses but I ran a hyperparameter optimizer (Optuna) over them to be more fact based. Of the 6 constants in this file, it changed `depthProbeInterval` - reducing it to 4. That did indeed produce better results overall by allowing the depth controller to recover faster when switching between different types of inputs.

File: x/mlxrunner/speculate.go (hunk 11 above)
Comment (jessegross) [reply x/mlxrunner/speculate.go:168]: Yes, that's a good catch. I changed it so that on the next round we just pick up the same depth that was never used when the last round ended rather than calling `depth.next` a second time in `open`.
